data-model/src/main/java/org/hypertrace/core/datamodel/shared/StructuredTraceGraph.java

```diff
@@ -48,40 +46,61 @@ private StructuredTraceGraph() {
    */
   public static StructuredTraceGraph createGraph(StructuredTrace trace) {
     StructuredTraceGraph graph = new StructuredTraceGraph();
+    graph.buildEventMap(trace);
+    graph.buildEntityMap(trace);
     graph.buildParentChildRelationship(trace);
     graph.buildRootEvents(trace);
     graph.buildRootEntities(trace);
     return graph;
   }
 
+  /**
+   * @return an immutable set containing the root events
+   */
   public Set<Event> getRootEvents() {
-    return ImmutableSet.copyOf(rootEvents);
+    return rootEvents;
   }
 
+  /**
+   * @return an immutable set containing the root entities
+   */
   public Set<Entity> getRootEntities() {
-    return ImmutableSet.copyOf(rootEntities);
+    return rootEntities;
   }
 
   public Event getParentEvent(Event event) {
-    return childToParentEvents.get(event);
+    return childToParentEvents.get(event.getEventId());
   }
 
   public List<Entity> getParentEntities(Entity entity) {
-    return childToParentEntities.get(entity);
+    return childToParentEntities.get(entity.getEntityId());
   }
 
   public List<Event> getChildrenEvents(Event event) {
-    return parentToChildrenEvents.get(event);
+    return parentToChildrenEvents.get(event.getEventId());
   }
 
   public List<Entity> getChildrenEntities(Entity entity) {
-    return parentToChildrenEntities.get(entity);
+    return parentToChildrenEntities.get(entity.getEntityId());
   }
 
+  /**
+   * @return an immutable map of entity ids to entities
+   */
   public Map<String, Entity> getEntityMap() {
     return entityMap;
   }
 
+  private void buildEventMap(StructuredTrace trace) {
+    eventMap = trace.getEventList().stream()
+        .collect(Collectors.toUnmodifiableMap(Event::getEventId, Function.identity()));
+  }
+
+  private void buildEntityMap(StructuredTrace trace) {
+    entityMap = trace.getEntityList().stream()
+        .collect(Collectors.toUnmodifiableMap(Entity::getEntityId, Function.identity()));
```

Comment:
If the trace contains duplicate entities, this throws an error while building the map because it expects unique keys. This can be handled by providing a merging function(we can use the latest entity).

Reply:
PR for fixing this: https://github.com/hypertrace/data-model/pull/15